### ISSUE TITLE
Rule Change: Make no-await-in-loop allow the `await` keyword inside async generator functions

### ISSUE THREAD

**Issue by leonardoraele** [eslint/eslint#15784]
### What rule do you want to change?

no-await-in-loop

### What change to do you want to make?

Generate fewer warnings

### How do you think the change should be implemented?

A new default behavior

### Example code

```js
async function* iterateStream(sourceStream) {
	const reader = sourceStream.getReader();
	try {
		while (true) {
			const { value, done } = await reader.read();
			if (done) {
				break;
			} else {
				yield value;
			}
		}
	} catch(error) {
		reader.releaseLock();
	}
}
```

### What does the rule currently do for this code?

It complains about `await` keyword being used inside a loop.

### What will the rule do after it's changed?

It should not complain about the `await` keyword being used inside loops in an async generator function.

### Participation

- [ ] I am willing to submit a pull request to implement this change.

### Additional comments

For reference, MDN documentation on `for await` syntax and async iterators:
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/for-await...of#iterating_over_async_generators

**Comment by nzakas** [comment]
> It should not complain about the await keyword being used inside loops in an async generator function.

Why not?

**Comment by leonardoraele** [comment]
Because that's how they are supposed to work...

Asynchronous generators are meant to be used in situations where you might not know ahead of time the information you will yield, like you do with normal async functions. (i.e. you don't have all the promises to `await` yet, so you can't replace the loop for a promise aggregator like `Promise.all`)

In the example code I added, we `await` data as it becomes available in the stream. We can't use `Promise.all` because we don't have a promise to wait yet for future data; we can only `await` for each chunk at a time.

You can take a look at the MDN page I mentioned, it has a very similar example:
> For reference, MDN documentation on for await syntax and async iterators:
> https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/for-await...of#iterating_over_async_generators

**Comment by nzakas** [comment]
I’m having trouble understanding why an async generator would be considered different from an async function in general. An async generator doesn’t necessarily imply a loop will be used (it’s certainly not required) any more than an async function implies that a loop won’t be used.

**Comment by leonardoraele** [comment]
The `no-await-in-loop` rule exists to encourage users to take advantage of paralellization of tasks by using `Promise.all` instead of loops with `await`s. The rule is not appropriate to situations where tasks can't be paralellized, like the example I provided.

When I wrote this suggestion, I assumed no paralellization was the usual case for async generator functions, but now that I put more thought into it, I realize you can also have loops that can't be paralellized in normal async functions, not only generators.

A way eslint could detect the impossibility of parallelizing tasks is by looking for loops whose exit condition depend on results of asynchronous tasks.

e.g.
```js
async function fop() {
  let loopContinues;
  while (loopContinues) {
    // The return of an asynchronous task is assigned to a variable that is part of the loop condition
    loopContinues = await somethig();
  }
}
```

In these cases, I think it is safe to assume the loop can't be paralellized, thus `no-await-in-loop` would not apply.

I have no idea how difficult would be to implement this into the rule.

**Comment by mdjermanovic** [comment]
> When I wrote this suggestion, I assumed no paralellization was the usual case for async generator functions

I think this is a valid assumption for the use of yielded values, but an async generator might still contain some loops that could be replaced with `Promise.all()`; for example, when a loop inside the generator is used to `await` multiple resources needed to calculate one yielded value.

Maybe the rule could check if there is a `yield` inside the loop? If it is, then the consumer of the generator controls when the next iteration will be performed and whether it is necessary to continue the work at all. In that case, disallowing `await` in favor of `Promise.all()`, which is the purpose of this rule, doesn't seem right as it suggests loading resources that might not be needed at a certain point or at all.

> A way eslint could detect the impossibility of parallelizing tasks is by looking for loops whose exit condition depend on results of asynchronous tasks.

That would be very difficult to detect reliably. Documentation for this rule ([when not to use it](https://eslint.org/docs/rules/no-await-in-loop#when-not-to-use-it)) suggests using ESLint disable comments in such cases.

**Comment by nzakas** [comment]
> Maybe the rule could check if there is a yield inside the loop?

The trick here is the relationship between await and yield. If you are awaiting something that is ultimately yielded, then you would not want a warning, otherwise you would want the current behavior. This seems a bit complex to me and I’m not sure it’s worth the complexity for this edge case.

**Comment by nzakas** [comment]
I’m 👎 to making a change here. Unless someone else wants to champion a change, we should close this issue.
